test/unit_test/src/test_Conv2dRegression_DW.cpp

```diff
@@ -0,0 +1,430 @@
+
+                            int8_t kernel_pad_val = rng.rand<int8_t>();
+
+                            std::array<int, 4> shape = {
+                                {1, k_height, k_width, x_channels}};
+
+                            Conv2dReorderedWeights rw =
+                                MatMulDirectFn_DW::reorder_kernel_weights(
+                                    (int8_t *)weights.data(), shape,
+                                    kernel_pad_val);
+
+                            MatMulDirectFn_DW::Params p(X, K, rw.weights.data(),
+                                                        (int)rw.weights.size());
+                            MatMulDirectFn_DW aggregator(&p);
+
+                            OutputTransformFnInt8::CanonicalMulAndBias
+                                canonical_values = OutputTransformFnInt8::
+                                    canonicalise_mul_and_bias_dw(
+                                        eff_mult, bias, weights, shape,
+                                        ks.input_zero_point,
+                                        ks.output_zero_point, x_channels);
+
+                            QuantisationParams qp =
+                                OutputTransformFnInt8::quantise_activation(
+                                    canonical_values.f_multipliers,
+                                    canonical_values.f_biases,
+                                    canonical_values.accu_min,
+                                    canonical_values.accu_max);
+
+                            // pad q.biases and  q.multipliers to a multiple
+                            // of VPU_INT16_EPV this is to work around array
+                            // over reads
+                            int16_t pad_val =
+                                rng.rand<int8_t>();  // this is arbitrary
+                            OutputTransformFn::pad(qp.biases, VPU_INT16_EPV,
+                                                   pad_val);
+                            OutputTransformFn::pad(qp.multipliers,
+                                                   (int)VPU_INT16_EPV, pad_val);
+
+                            OT_int8::Params ot_params((int32_t)x_channels,
+                                                      &qp.otv, qp.biases,
+                                                      qp.multipliers);
+                            OT_int8 ot(&ot_params);
+                            auto ir = ImageRegion(0, 0, 0, Y.height, Y.width,
+                                                  Y.depth);
+
+                            Filter2D_DW::Params akp(Y, ir, VPU_INT8_ACC_PERIOD);
+
+                            Conv2dDepthwiseValidDirect conv2d(&akp, &memcpy,
+                                                              &aggregator, &ot);
+
+                            auto output = std::vector<int8_t>(
+                                Y.height * Y.width * Y.depth);
+
+                            conv2d.execute(&output[0], &input[0]);
```

Comment:
Do we need scratch for DWValidDirect?

Reply:
No, the direct nature is that it operates directly on the input tensor

Reply:
No, the direct nature of the kernel is that it operated directly in the input tensor

Reply:
That's interesting. I think I will need to fix this section in the compiler. Scratch is calculated for all types, https://github.com/xmos/ai_tools/blob/develop/experimental/xformer/Transforms/ReplaceWithConv2DV2.cpp#L457